src/CodeGenerators/SettingsGen/BaseGenerator.cs

```diff
@@ -0,0 +1,130 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT license.
+
+using System;
+using System.Collections.Generic;
+using Microsoft.CodeAnalysis;
+using Microsoft.Omex.CodeGenerators.SettingsGen.FileGeneration;
+using Microsoft.Omex.CodeGenerators.SettingsGen.Parser;
+using Microsoft.Omex.CodeGenerators.SettingsGen.Validation;
+
+namespace Microsoft.Omex.CodeGenerators.SettingsGen
+{
+	/// <summary>
+	/// The base generator which can be inherited in order to implement your own version
+	/// </summary>
+	/// <typeparam name="TSettingModel">Settings model type</typeparam>
+	public abstract class BaseGenerator<TSettingModel> : ISourceGenerator
+		where TSettingModel : class, IEqualityComparer<TSettingModel>, IEquatable<TSettingModel>
+	{
+		/// <inheritdoc />
+		public void Execute(GeneratorExecutionContext context)
+		{
+			if (context.SyntaxContextReceiver is not IParser<TSettingModel> parser)
+			{
+				return;
+			}
+
+			try
+			{
+				TSettingModel settings = parser.GetSettings();
+				bool shouldGen = ShouldGenerateFile(context, out AdditionalText? settingsFile);
+
+				if (!shouldGen)
+				{
+					context.ReportDiagnostic(Diagnostic.Create(s_dontGen, null));
+					return;
+				}
+
+				if (settingsFile == null)
+				{
+					throw new Exception("No file to compare or generate given");
+				}
+
+				if (Validator?.AreExistingSettingsEqual(settings, settingsFile) == true)
+				{
+					context.ReportDiagnostic(Diagnostic.Create(s_matchingSettings, null));
+					return;
+				}
+
+				if (Filegenerator is null)
+				{
+					throw new Exception($"No file generator set for class {GetType().Name}");
+				}
+
+				context.ReportDiagnostic(Diagnostic.Create(s_writingToFile, null, settingsFile.Path));
+
+				Filegenerator.GenerateFile(settings, settingsFile.Path);
+			}
+			catch (Exception ex)
+			{
+				context.ReportDiagnostic(Diagnostic.Create(s_failedGeneration, null, ex.Message));
+				throw;
+			}
+		}
+
+		/// <inheritdoc />
+		public void Initialize(GeneratorInitializationContext context)
+		{
+			if (Parser != null)
```

Comment:
What happens if parser is null? I think it would not matter so you can just register without the null check?